### PR TITLE
secretmanager: add initial samples

### PR DESCRIPTION
**Pull request by sethvargo**
I also noticed there's some fun Koroko scripts that download secrets from GCS. Let me know if you'd like some help to migrate those to Secret Manager 😉

### CODE REVIEW

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+            SecretManagerServiceClient client = SecretManagerServiceClient.Create();
+
+            // Create the request.
+            var request = new UpdateSecretRequest
+            {
+                Secret = new Secret
+                {
+                  Name = name,
+                  // Labels = new Dictionary<string, string>
```
Comment (sethvargo) [inline secretmanager/api/SecretManagerSample/Program.cs]: Help/guidance here would be appreciated. Labels is readonly, so I'm confused how to update them. Is that a bug in the generated library?

Comment (jskeet) [reply secretmanager/api/SecretManagerSample/Program.cs]: No - the property is readonly, but it returns a mutable map. See https://googleapis.github.io/google-cloud-dotnet/docs/faq.html#how-can-i-modify-repeated-fields-and-maps-in-protobuf-messages

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+using Google.Protobuf.WellKnownTypes;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
+
+namespace GoogleCloudSamples
+{
+
+    class SecretOptions {
+        [Value(0, HelpText = "Full resource name of the secret", Required = true)]
+        public string name { get; set; }
```
Comment (amanda-tarafa) [inline secretmanager/api/SecretManagerSample/Program.cs]: nit: This are better named with a capital N. If you are editing a file that already has some options named with first lowercase, then it's allright to keep to that, but since this is a new file, please, name all CamelCase.

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+            var version = client.AccessSecretVersion(request);
+            string payload = version.Payload.Data.ToStringUtf8();
+            Console.WriteLine($"Payload: {payload}");
+        }
+        // [END secretmanager_access_secret_version]
+
+        // [START secretmanager_add_secret_version]
+        public static void AddSecretVersion(string parent)
+        {
+            // string parent = "projects/my-project/secrets/my-secret";
```
Comment (amanda-tarafa) [inline secretmanager/api/SecretManagerSample/Program.cs]: Instead of using this commented line here (if uncommented, this won't compile), prefer using parameters with default values or better still XML documentation for the method.

`public static void AddSecretVersion(string parent = "projects/my-project/secrets/my-secret")`

```
/// <summary>
/// ...
/// </summary>
/// <param name="parent">....</param>
```

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            // Add the secret version.
+            var version = client.AddSecretVersion(request);
+            Console.WriteLine($"Added secret version {version.Name}");
+        }
+        // [END secretmanager_add_secret_version]
+
+        // [START secretmanager_create_secret]
+        public static void CreateSecret(string parent, string id)
+        {
+            // string parent = "projects/my-project";
```
Comment (amanda-tarafa) [inline secretmanager/api/SecretManagerSample/Program.cs]: Same as before, and in general for all others.

**Comment by amanda-tarafa** [comment]
Also, CI is broken partly because of some dependency updates. I'm fixing that and plan to have that fix merged in a few hours. When you are done with the changes plese check that all your tests pass on CI.
As for updating the Kokoro scripts, I don't have a final answer to whether we need migrating those or not, but if you'd like to start another PR for that we can work from there. Your help would be really appreciated ;).

**Comment by sethvargo** [comment]
Hey @amanda-tarafa - I've made the requested updates. Once we get the samples reviewed and merged across all the languages, I'll take a pass at any libraries that have secrets and look into an update 😄

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+        /// Accesses a secret with provided version.
+        /// </summary>
+        /// <param name="name">Secret version name to access.</param>
+        /// <example>
+        /// With a specific version.
+        /// <code>AccessSecretVersion("projects/my-project/secrets/my-secret/versions/5")</code>
+        /// </example>
+        /// <example>
+        /// With an alias version.
+        /// <code>AccessSecretVersion("projects/my-project/secrets/my-secret/versions/latest")</code>
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/SecretManagerSample/Program.cs]: Done

**Comment by sethvargo** [comment]
Hi @amanda-tarafa please re-review. There are unrelated (Leaderboard) tests failing. I don't see any tests related to secrets failing.

**Comment by sethvargo** [comment]
> Secret Manager API has not been used in project 1061780273775 before or it is disabled...

Can someone with permissions run `gcloud services enable secretmanager.googleapis.com` on the appropriate test projects please? @amanda-tarafa

**Comment by jskeet** [comment]
Unfortunately I don't have permission to do this :( Amanda is back from vacation on Monday.

**Comment by amanda-tarafa** [comment]
I also don't have permissions to enable APIs. @SurferJeffAtGoogle can you please take a look.

**Comment by amanda-tarafa** [comment]
@sethvargo I'm shortly pushing a PR that will fix the Translate test that is failing.
The Logging test that is timing out is known flakiness, so that you can ignore, if it happens again, rerun CI by adding the kokoro:force-run label to the PR until is green. I'll fix the test sometime in the near future.
Also, I've asked @SurferJeffAtGoogle for ownership of the test project so I can enable the API.
Let me know if there are more issues you are running into.
When all is green let me know and I can take another look at the PR.

**Comment by sethvargo** [comment]
Hi @amanda-tarafa - I believe it's currently green for secret manager. From the build output

```text
Starting .\secretmanager\api\SecretManagerTest\runTests.ps1...
-------------------------------------------------------------------------------
.\secretmanager\api\SecretManagerTest\runTests.ps1
  Restore completed in 2.09 sec for T:\src\github\dotnet-docs-samples\testutil\testutil.csproj.
  Restore completed in 11.76 sec for T:\src\github\dotnet-docs-samples\secretmanager\api\QuickStart\QuickStart.csproj.
  Restore completed in 11.88 sec for T:\src\github\dotnet-docs-samples\secretmanager\api\SecretManagerSample\SecretManagerSample.csproj.
  Restore completed in 12.11 sec for T:\src\github\dotnet-docs-samples\secretmanager\api\SecretManagerTest\SecretManagerTest.csproj.
Test run for T:\src\github\dotnet-docs-samples\secretmanager\api\SecretManagerTest\bin\Debug\netcoreapp2.0\SecretManagerTest.dll(.NETCoreApp,Version=v2.0)
Microsoft (R) Test Execution Command Line Tool Version 16.0.1
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
SecretManagerSample create dotnet-docs-samples-tests csharp-20200109190052640
QuickStart dotnet-docs-samples-tests csharp-quickstart-20200109190052640-1
SecretManagerSample add-version dotnet-docs-samples-tests csharp-20200109190052640
SecretManagerSample access-version dotnet-docs-samples-tests csharp-20200109190052640 1
SecretManagerSample delete dotnet-docs-samples-tests csharp-20200109190052640
SecretManagerSample create dotnet-docs-samples-tests csharp-20200109190054468
SecretManagerSample add-version dotnet-docs-samples-tests csharp-20200109190054468
SecretManagerSample disable-version dotnet-docs-samples-tests csharp-20200109190054468 1
SecretManagerSample enable-version dotnet-docs-samples-tests csharp-20200109190054468 1
SecretManagerSample destroy-version dotnet-docs-samples-tests csharp-20200109190054468 1
SecretManagerSample delete dotnet-docs-samples-tests csharp-20200109190054468
SecretManagerSample create dotnet-docs-samples-tests csharp-list-20200109190055250-1
SecretManagerSample create dotnet-docs-samples-tests csharp-list-20200109190055250-2
SecretManagerSample create dotnet-docs-samples-tests csharp-list-20200109190055250-3
SecretManagerSample add-version dotnet-docs-samples-tests csharp-list-20200109190055250-1
SecretManagerSample add-version dotnet-docs-samples-tests csharp-list-20200109190055250-1
SecretManagerSample add-version dotnet-docs-samples-tests csharp-list-20200109190055250-1
SecretManagerSample list dotnet-docs-samples-tests
SecretManagerSample list-versions dotnet-docs-samples-tests csharp-list-20200109190055250-1
SecretManagerSample delete dotnet-docs-samples-tests csharp-list-20200109190055250-1
SecretManagerSample delete dotnet-docs-samples-tests csharp-list-20200109190055250-2
SecretManagerSample delete dotnet-docs-samples-tests csharp-list-20200109190055250-3
Results File: T:\src\github\dotnet-docs-samples\secretmanager\api\SecretManagerTest\TestResults\TestResults.xml

Total tests: 4. Passed: 4. Failed: 0. Skipped: 0.
Test Run Successful.
Test execution time: 8.0770 Seconds
```

File: secretmanager/api/SecretManagerTest/SecretManagerTest.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        public SampleTests()
+        {
+            if (s_projectId == null || s_projectId == "")
+            {
+                throw new Exception("missing GOOGLE_PROJECT_ID");
+            }
+        }
+
+        [Fact]
+        public void TestCreateAddAccessDelete()
```
Comment (amanda-tarafa) [inline secretmanager/api/SecretManagerTest/SecretManagerTest.cs]: I'm a little worried about the maintenability of these tests that test multiple things. I understand that for testing delete you first need to create, etc. but I'd prefer a test that eclusively tests create, another delete, etc.
For doing that, consider creating a [Fixture](https://xunit.net/docs/shared-context#class-fixture) that can hold all the methods you need for setting up your tests and also for instance, a `void MarkForDeletion(Secret secret)` that saves a collection of secrets to delete, that are deleted on dispose. Take a look at this [StorageFixture](https://github.com/googleapis/google-cloud-dotnet/blob/master/apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/StorageFixture.cs) for instance. Let me know if I can help with that.

File: secretmanager/api/QuickStart/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+1.  Edit [QuickStart.cs](QuickStart/QuickStart.cs), and replace
+    `YOUR-PROJECT-ID` with the id of the project you created in step 1.
+
+1.  From a command line with `dotnet` available:
+
+    ```text
+    PS C:\...\dotnet-docs-samples\secretmanager\api\QuickStart> dotnet restore
+    PS C:\...\dotnet-docs-samples\secretmanager\api\QuickStart> dotnet run
+    Secrets:
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/QuickStart/README.md]: This is the output of the command. I'll make it clearer by moving it to it's own block.

File: secretmanager/api/QuickStart/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+## Build and Run
+
+1.  **Follow the set-up instructions in [the documentation](https://cloud.google.com/dotnet/docs/setup).**
+
+1.  Enable APIs for your project.
+    [Click here](https://console.cloud.google.com/flows/enableapi?apiid=secretmanager.googleapis.com&showconfirmation=true)
+    to visit Cloud Platform Console and enable the Secret Manager API.
+
+1.  Edit [QuickStart.cs](QuickStart/QuickStart.cs), and replace
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/QuickStart/README.md]: Good catch. I'm now guarding that with a conditional. It's not just project, it's also secret id. The tests randomly generate a UUID (because secret names have to be unique) and also cleans them up afterward. The only other option I can think of is an environment variable, but I think this is a lot cleaner (and what we did in the samples of the other languages).

File: secretmanager/api/QuickStart/QuickStart.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        public static void Main(string[] args)
+        {
+            // GCP project in which to store secrets in Secret Manager.
+            string projectId = "YOUR-PROJECT-ID";
+
+            // ID of the secret to create.
+            string secretId = "YOUR-SECRET-ID";
+
+            // [END secretmanager_quickstart]
+            projectId = args[0];
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/QuickStart/QuickStart.cs]: I commented above, but it makes testing really difficult otherwise, since secret-id is a UUID that we also need to cleanup.

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+using System.Text;
+
+using CommandLine;
+using Google.Api.Gax.ResourceNames;
+using Google.Cloud.Iam.V1;
+using Google.Cloud.SecretManager.V1Beta1;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/SecretManagerSample/Program.cs]: Done

File: secretmanager/api/SecretManagerSample/Program.cs [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+        /// <param name="secretId">ID of the secret.</param>
+        /// <example>
+        /// List all secret versions.
+        /// <code>ListSecretVersions("my-project", "my-secret")</code>
+        /// </example>
+        public static void ListSecretVersions(string projectId, string secretId)
+        {
+            SecretManagerServiceClient client = SecretManagerServiceClient.Create();
+
+            Console.WriteLine($"projectId: ${projectId}");
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/SecretManagerSample/Program.cs]: That's leftover debug code - good catch. Removed.

File: secretmanager/api/QuickStart/README.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+A sample that demonstrates how to call the
+[Secret Manager API](https://cloud.google.com/secret-manager/docs) from C#.
+
+This sample requires [.NET Core 2.1](
+    https://www.microsoft.com/net/core) or later.
+
+## Build and Run
+
+1.  **Follow the set-up instructions in [the documentation](https://cloud.google.com/dotnet/docs/setup).**
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/QuickStart/README.md]: Yup - that's totally valid markdown (and markdown will do the right numbering when rendered 😄). It's a huge timesaver instead of trying to rename and shift numbers.

In fact, you can also do something like this:

```
1.  a
7.  b
2.  c
1005.  d
```

and it will render:

1.  a
7.  b
2.  c
1005.  d

File: secretmanager/api/QuickStart/README.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    [Click here](https://console.cloud.google.com/flows/enableapi?apiid=secretmanager.googleapis.com&showconfirmation=true)
+    to visit Cloud Platform Console and enable the Secret Manager API.
+
+1.  Edit [QuickStart.cs](QuickStart/QuickStart.cs), and replace
+    `YOUR-PROJECT-ID` with the id of the project you created in step 1.
+
+1.  From a command line with `dotnet` available:
+
+    ```text
+    PS C:\...\dotnet-docs-samples\secretmanager\api\QuickStart> dotnet restore
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/QuickStart/README.md]: Technically no? I copied this from all the other examples though. Do we want SM to be different from all the other samples?

File: secretmanager/api/SecretManagerSample/README.md [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+[Secret Manager API](https://cloud.google.com/secret-manager) from C#.
+
+This sample requires [.NET Core 3.0](
+    https://www.microsoft.com/net/core) or later.  That means using
+[Visual Studio 2017](
+    https://www.visualstudio.com/), or the command line.
+
+## Build and Run
+
+1.  **Follow the set-up instructions in [the documentation](https://cloud.google.com/dotnet/docs/setup).**
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply secretmanager/api/SecretManagerSample/README.md]: You caught me copy-pasting. Fixed to all be 1.s

File: secretmanager/api/SecretManagerTest/SecretManagerTest.cs (hunk 6 above)
Comment (sethvargo) [reply secretmanager/api/SecretManagerTest/SecretManagerTest.cs]: A challenge with the other language samples was quotas. It depends on how fast the tests execute. Let me rewrite them and see if they aren't quota-ed out.

Comment (sethvargo) [reply secretmanager/api/SecretManagerTest/SecretManagerTest.cs]: A challenge with the other language samples was quotas. It depends on how fast the tests execute. Let me rewrite them and see if they aren't quota-ed out.

Comment (sethvargo) [reply secretmanager/api/SecretManagerTest/SecretManagerTest.cs]: Okay I just pushed up some commits that break the tests apart.

**Comment by sethvargo** [comment]
Hey @amanda-tarafa I believe I've addressed all your comments. PTAL

File: secretmanager/api/SecretManagerTest/SecretsFixture.cs [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+namespace GoogleCloudSamples
+{
+    /// <summary>
+    /// Fixture for secrets used in all tests.
+    /// <summary>
+    public class SecretsFixture : IDisposable
+    {
+        private static readonly string s_projectId = Environment.GetEnvironmentVariable("GOOGLE_PROJECT_ID");
+        private static readonly SecretManagerServiceClient s_client = SecretManagerServiceClient.Create();
```
Comment (amanda-tarafa) [inline secretmanager/api/SecretManagerTest/SecretsFixture.cs]: The client better not be static.

File: secretmanager/api/SecretManagerTest/SecretsFixture.cs [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+        public SecretVersion SecretVersionToEnable { get; private set; }
+
+        public SecretsFixture()
+        {
+            if (String.IsNullOrEmpty(s_projectId))
+            {
+                throw new Exception("missing GOOGLE_PROJECT_ID");
+            }
+
+            Secret = CreateSecret();
```
Comment (amanda-tarafa) [inline secretmanager/api/SecretManagerTest/SecretsFixture.cs]: If quotas are an issue for managing secrets in the fixture, we can add a tiny sleep after each create/delete/modify operation. But I'm happy for that to be done in a separate PR if we have trouble.

File: secretmanager/api/QuickStart/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply secretmanager/api/QuickStart/README.md]: OK, leave it as is for consistency. At some point we'll rehaul all of this, but that's in the future.
